oar/core/shipment.py

```diff
@@ -782,3 +784,189 @@ def drop_bugs(self) -> tuple[list[str], list[str]]:
                 raise ShipmentDataException(f"Failed to drop bugs: {str(e)}")
         
         return all_high_severity_bugs, all_can_drop_issues
+
+    def _get_components_from_shipment(self, mr: GitLabMergeRequest) -> list[dict]:
+        """Get all components from shipment YAML files
+        
+        Args:
+            mr: GitLabMergeRequest instance to get files from
+            
+        Returns:
+            List of component objects from shipment.snapshot.spec.components
+            
+        Raises:
+            ShipmentDataException: If unable to process shipment files
+        """
+        components = []
+        try:
+            yaml_files = mr.get_all_files()
+            for file_path in yaml_files:
+                try:
+                    content = mr.get_file_content(file_path)
+                    data = yaml.safe_load(content)
+                    components.extend(glom(data, 'shipment.snapshot.spec.components', default=[]))
+                except Exception as e:
+                    logger.warning(f"Failed to process file {file_path}: {str(e)}")
+                    continue
+        except Exception as e:
+            raise ShipmentDataException(f"Failed to get components from shipment: {str(e)}")
+        return components
+
+    def _get_image_digest(self, pull_spec: str) -> str:
+        """Extract image digest from container image pull spec
+        
+        Args:
+            pull_spec: Container image pull spec string
+            
+        Returns:
+            Image digest (sha256:...)
+            
+        Raises:
+            ShipmentDataException: If pull spec is invalid or missing digest
+        """
+        if not pull_spec:
+            raise ShipmentDataException("Empty pull spec provided")
+            
+        if "@sha256:" not in pull_spec:
+            raise ShipmentDataException(f"Pull spec {pull_spec} does not contain digest")
+            
+        return pull_spec.split("@sha256:")[1]
+
+    def _query_pyxis_freshness(self, image_digest: str) -> list[dict]:
+        """Query Pyxis API for image freshness grades
+        
+        Args:
+            image_digest: Image digest to query (sha256:...)
+            
+        Returns:
+            List of freshness grade objects from Pyxis API
+            
+        Raises:
+            ShipmentDataException: If API request fails
+        """
+        
+        try:
+            url = f"https://catalog.stage.redhat.com/api/containers/v1/images?filter=image_id==sha256:{image_digest}&page_size=100&page=0"
+            proxies = {"https": "squid.corp.redhat.com:3128"}
+            
+            response = requests.get(url, proxies=proxies)
+            response.raise_for_status()
+            
+            data = response.json()
+            if not data.get("data"):
+                return []
+            return data["data"][0].get("freshness_grades", [])
```

Comment:
What is the first item `[0]`?  I am not able to access the url. I see `Preprod Lockdown: Access Denied`. Do I need some permission?

Reply:
you need to access the endpoint with proxy